autodispose/src/main/java/com/uber/autodispose/AutoDisposingMaybeObserverImpl.java

```diff
@@ -41,24 +41,29 @@
   }
 
   @Override public void onSubscribe(final Disposable d) {
-    if (AutoDisposeEndConsumerHelper.setOnce(lifecycleDisposable,
-        lifecycle.subscribeWith(new DisposableMaybeObserver<Object>() {
-          @Override public void onSuccess(Object o) {
-            callMainSubscribeIfNecessary(d);
-            AutoDisposingMaybeObserverImpl.this.dispose();
-          }
+    DisposableMaybeObserver<Object> o = new DisposableMaybeObserver<Object>() {
+      @Override public void onSuccess(Object o) {
+        callMainSubscribeIfNecessary(d);
+        AutoDisposableHelper.dispose(mainDisposable);
+        lifecycleDisposable.lazySet(AutoDisposableHelper.DISPOSED);
+      }
 
-          @Override public void onError(Throwable e) {
-            callMainSubscribeIfNecessary(d);
-            AutoDisposingMaybeObserverImpl.this.onError(e);
-          }
+      @Override public void onError(Throwable e) {
+        callMainSubscribeIfNecessary(d);
+        AutoDisposingMaybeObserverImpl.this.onError(e);
+        mainDisposable.lazySet(AutoDisposableHelper.DISPOSED);
+        lifecycleDisposable.lazySet(AutoDisposableHelper.DISPOSED);
+      }
 
-          @Override public void onComplete() {
-            callMainSubscribeIfNecessary(d);
-            // Noop - we're unbound now
-          }
-        }),
-        getClass())) {
+      @Override public void onComplete() {
+        callMainSubscribeIfNecessary(d);
+        mainDisposable.lazySet(AutoDisposableHelper.DISPOSED);
+        lifecycleDisposable.lazySet(AutoDisposableHelper.DISPOSED);
+        // Noop - we're unbound now
+      }
+    };
+    if (AutoDisposeEndConsumerHelper.setOnce(lifecycleDisposable, o, getClass())) {
+      lifecycle.subscribe(o);
       if (AutoDisposeEndConsumerHelper.setOnce(mainDisposable, d, getClass())) {
         delegate.onSubscribe(this);
```

Comment:
`callMainSubscribeIfNecessary` may cause double `onSubscribe` here.

Reply:
part of ba8bd87